### PR TITLE
Closes #1432 Remove Serializer's FileBasedPersistentContext constructor contract

### LINKED ISSUE #1432: Remove initial Serializer constructor contract with FileBasedPersistentContext
Following the work on #1411 we no longer rely on constructor specifics to indicate stateful serializers. And since we end up breaking the contract introduced in 3.1 when releasing 3.2, let's go all the way and remove the 3.0 contract which no longer makes sense.

### PR DESCRIPTION
**Pull request by albinsuresh**

### CODE REVIEW

File: docs/src/docs/asciidoc/user/serializers-copiers.adoc [hunk 1]
```diff
@@ -159,7 +159,7 @@ For persistent caches it is highly recommended that all state is stored in these
 * For clustered caches the contents are persisted in the cluster itself so that other clients using the same cache can also access the contents of the map.
 
 NOTE: The constructor with the signature `(ClassLoader classLoader, FileBasedPersistenceContext persistenceContext)`
-      that existed in v3.0 is still supported to respect backward compatibility but the usage is limited to disk based caches.
+      that existed till v3.1 has been deprecated since v3.2.
```
Comment (ljacomet) [inline docs/src/docs/asciidoc/user/serializers-copiers.adoc]: Not deprecated - removed 😉

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 2]
```diff
@@ -64,31 +64,25 @@
...
-      transientProvider = new TransientProvider(Collections.<Class<?>, Class<? extends Serializer<?>>>emptyMap());
-      persistentProvider = new PersistentProvider(Collections.<Class<?>, Class<? extends Serializer<?>>>emptyMap());
+      this.serializers = new LinkedHashMap<Class<?>, Class<? extends Serializer<?>>>(Collections.<Class<?>, Class<? extends Serializer<?>>>emptyMap());
     }
   }
 
   @Override
   public <T> Serializer<T> createKeySerializer(Class<T> clazz, ClassLoader classLoader, ServiceConfiguration<?>... configs) throws UnsupportedTypeException {
-    Serializer<T> serializer = (Serializer<T>)getUserProvidedSerializer(find(DefaultSerializerConfiguration.Type.KEY, configs));
+    Serializer<T> serializer = (Serializer<T>) getUserProvidedSerializer(find(DefaultSerializerConfiguration.Type.KEY, configs));
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: Replacing this line with:
```java
DefaultSerializerConfiguration<T> configuration = find(DefaultSerializerConfiguration.Type.KEY, configs);
Serializer<T> serializer = getUserProvidedSerializer(configuration);
```
will resolve generics warnings.

Also applies to `createValueSerializer`

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 3]
```diff
@@ -97,19 +91,71 @@ public DefaultSerializationProvider(DefaultSerializationProviderConfiguration co
...
+
+  protected <T> Class<? extends Serializer<T>> getSerializerClassFor(Class<T> clazz, DefaultSerializerConfiguration<T> config, ClassLoader classLoader) throws UnsupportedTypeException {
+    if (config != null) {
+      Class<? extends Serializer<T>> configured = config.getClazz();
+      if (configured != null) {
+        return configured;
+      }
+    }
+
+    Class<? extends Serializer<T>> direct = (Class<? extends Serializer<T>>) serializers.get(clazz);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: Needs `@SuppressWarnings("unchecked")`

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 4]
```diff
@@ -97,19 +91,71 @@ public DefaultSerializationProvider(DefaultSerializationProviderConfiguration co
...
+    if (direct != null) {
+      return direct;
+    }
+    for (Map.Entry<Class<?>, Class<? extends Serializer<?>>> entry : serializers.entrySet()) {
+      if (entry.getKey().isAssignableFrom(clazz)) {
+        return (Class<? extends Serializer<T>>) entry.getValue();
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: Needs local var that can carry `@SuppressWarnings("unchecked")` before return

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 5]
```diff
@@ -138,8 +184,14 @@ public void releaseSerializer(final Serializer<?> serializer) throws IOException
 
   @Override
   public void start(ServiceProvider<Service> serviceProvider) {
-    transientProvider.start(serviceProvider);
-    persistentProvider.start(serviceProvider);
+    addDefaultSerializerIfNoneRegistered(serializers, Serializable.class, (Class) CompactJavaSerializer.class);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: Needs a trick similar to what's done [here](https://github.com/ehcache/ehcache3/pull/1431/commits/cef191af7c1f4614244cb71900066117e68aac0e)

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 6]
```diff
@@ -64,31 +64,25 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(DefaultSerializationProvider.class);
 
-  private final TransientProvider transientProvider;
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: Deletes orphaned a number of import statements - please remove them

File: impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java [hunk 7]
```diff
@@ -97,19 +91,74 @@ public DefaultSerializationProvider(DefaultSerializationProviderConfiguration co
...
+
+    try {
+      return constructSerializer(clazz, klazz.getConstructor(ClassLoader.class), classLoader);
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException(klazz + " does not have a constructor that takes in a ClassLoader.", e);
+    }
+  }
+
+
+  private <T> Class<? extends Serializer<T>> getSerializerClassFor(Class<T> clazz, DefaultSerializerConfiguration<T> config, ClassLoader classLoader) throws UnsupportedTypeException {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/spi/serialization/DefaultSerializationProvider.java]: `classLoader` param is never used

File: impl/src/main/java/org/ehcache/impl/config/serializer/DefaultSerializationProviderConfiguration.java [hunk 8]
```diff
@@ -24,14 +24,17 @@
 import org.ehcache.spi.serialization.Serializer;
 import org.ehcache.core.spi.service.FileBasedPersistenceContext;
 import org.ehcache.spi.service.ServiceCreationConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/config/serializer/DefaultSerializationProviderConfiguration.java]: Deletes also created orphaned imports - please remove

File: impl/src/test/java/org/ehcache/impl/internal/store/heap/OnHeapStoreByValueTest.java [hunk 9]
```diff
@@ -73,7 +73,7 @@ public Long copyForWrite(Long obj) {
   public void testKeyCopierCalledOnGetOrComputeIfAbsent() throws Exception {
     LongCopier keyCopier = new LongCopier();
     OnHeapStore<Long, Long> store = newStore(SystemTimeSource.INSTANCE, Expirations.noExpiration(), Eviction.noAdvice(),
-        keyCopier, new SerializingCopier<Long>(new CompactJavaSerializer<Long>(ClassLoader.getSystemClassLoader())), 100);
+        keyCopier, new SerializingCopier<Long>(new JavaSerializer<Long>(ClassLoader.getSystemClassLoader())), 100);
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/impl/internal/store/heap/OnHeapStoreByValueTest.java]: Creates an orphaned import statement - please remove

**Comment by albinsuresh** [comment]
Addressed the review comments.

**Pull request merged**